File: src/main/scala/com/microsoft/ml/spark/opencv/ImageTransformer.scala
// Copyright (C) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in project root for information.

package com.microsoft.ml.spark.opencv

import com.microsoft.ml.spark.codegen.Wrappable
import com.microsoft.ml.spark.core.contracts.{HasInputCol, HasOutputCol}
import com.microsoft.ml.spark.core.schema.{BinaryFileSchema, ImageSchemaUtils}
import com.microsoft.ml.spark.logging.BasicLogging
import org.apache.spark.injections.UDFUtils
import org.apache.spark.ml.image.ImageSchema
import org.apache.spark.ml.param.{ParamMap, _}
import org.apache.spark.ml.util.{DefaultParamsReadable, DefaultParamsWritable, Identifiable}
import org.apache.spark.ml.{ImageInjections, Transformer}
import org.apache.spark.sql.types._
import org.apache.spark.sql.{DataFrame, Dataset, Row}
import org.opencv.core.{Core, Mat, Rect, Size}
import org.opencv.imgproc.Imgproc

import scala.collection.mutable.ListBuffer

//scalastyle:off field.name
/** Image processing stage.
  *
  * @param params Map of parameters
  */
abstract class ImageTransformerStage(params: Map[String, Any]) extends Serializable {
  def apply(image: Mat): Mat

  val stageName: String
}

/** Resizes the image. The parameters of the ParameterMap are:
  * "height" - the height of the image
  * "width"
  * "stageName"
  * Please refer to [[http://docs.opencv.org/2.4/modules/imgproc/doc/geometric_transformations.html#resize OpenCV]]
  * for more information
  *
  * @param params ParameterMap of the parameters
  */
class ResizeImage(params: Map[String, Any]) extends ImageTransformerStage(params) {
  val height: Double = params(ResizeImage.height).asInstanceOf[Int].toDouble
  val width: Double = params(ResizeImage.width).asInstanceOf[Int].toDouble
  override val stageName: String = ResizeImage.stageName

  override def apply(image: Mat): Mat = {
    val resized = new Mat()
    val sz = new Size(width, height)
    Imgproc.resize(image, resized, sz)
    resized
  }
}

/** Resize object contains the information for resizing;
  * "height"
  * "width"
  * "stageName" = "resize"
  */
object ResizeImage {
  val stageName = "resize"
  val height = "height"
  val width = "width"
}

/** Crops the image for processing. The parameters are:
  * "x" - First dimension; start of crop
  * "y" - second dimension - start of crop
  * "height" -height of cropped image
  * "width" - width of cropped image
  * "stageName" - "crop"
  *
  * @param params ParameterMap of the dimensions for cropping
  */
class CropImage(params: Map[String, Any]) extends ImageTransformerStage(params) {
  val x: Int = params(CropImage.x).asInstanceOf[Int]
  val y: Int = params(CropImage.y).asInstanceOf[Int]
  val height: Int = params(CropImage.height).asInstanceOf[Int]
  val width: Int = params(CropImage.width).asInstanceOf[Int]
  override val stageName: String = CropImage.stageName

  override def apply(image: Mat): Mat = {
    val rect = new Rect(x, y, width, height)
    new Mat(image, rect)
  }
}

object CropImage {
  val stageName = "crop"
  val x = "x"
  val y = "y"
  val height = "height"
  val width = "width"
}

/** Converts an image from one color space to another, eg COLOR_BGR2GRAY. Refer to
  * [[http://docs.opencv.org/2.4/modules/imgproc/doc/miscellaneous_transformations.html#cvtcolor OpenCV]]
  * for more information.
  *
  * @param params Map of parameters and values
  */
class ColorFormat(params: Map[String, Any]) extends ImageTransformerStage(params) {
  val format: Int = params(ColorFormat.format).asInstanceOf[Int]
  override val stageName: String = ColorFormat.stageName

  override def apply(image: Mat): Mat = {
    val dst = new Mat()
    Imgproc.cvtColor(image, dst, format)
    dst
  }
}

object ColorFormat {
  val stageName = "colorformat"
  val format = "format"
}

/** Flips the image
  *
  * @param params
  */
class Flip(params: Map[String, Any]) extends ImageTransformerStage(params) {
  val flipCode = params(Flip.flipCode).asInstanceOf[Int]
  override val stageName = Flip.stageName

  override def apply(image: Mat): Mat = {
    val dst = new Mat()
    Core.flip(image, dst, flipCode)
    dst
  }
}

object Flip {
  val stageName: String = "flip"
  val flipCode: String = "flipCode"

  val flipUpDown: Int = 0
  val flipLeftRight: Int = 1
  val flipBoth: Int = -1
}

/** Blurs the image using a box filter.
  * The com.microsoft.ml.spark.core.serialize.params are a map of the dimensions of the blurring box. Please refer to
  * [[http://docs.opencv.org/2.4/modules/imgproc/doc/filtering.html#blur OpenCV]] for more information.
  *
  * @param params
  */
class Blur(params: Map[String, Any]) extends ImageTransformerStage(params) {
  val height: Double = params(Blur.height).asInstanceOf[Double]
  val width: Double = params(Blur.width).asInstanceOf[Double]
  override val stageName: String = Blur.stageName

  override def apply(image: Mat): Mat = {
    val dst = new Mat()
    Imgproc.blur(image, dst, new Size(height, width))
    dst
  }
}

object Blur {
  val stageName: String = "blur"
  val height: String = "height"
  val width: String = "width"
}

/** Applies a threshold to each element of the image. Please refer to
  * [[http://docs.opencv.org/2.4/modules/imgproc/doc/miscellaneous_transformations.html#threshold threshold]] for
  * more information
  *
  * @param params
  */
class Threshold(params: Map[String, Any]) extends ImageTransformerStage(params) {
  val threshold: Double = params(Threshold.threshold).asInstanceOf[Double]
  val maxVal: Double = params(Threshold.maxVal).asInstanceOf[Double]
  // EG Imgproc.THRESH_BINARY
  val thresholdType: Int = params(Threshold.thresholdType).asInstanceOf[Int]
  override val stageName: String = Threshold.stageName

  override def apply(image: Mat): Mat = {
    val dst = new Mat()
    Imgproc.threshold(image, dst, threshold, maxVal, thresholdType)
    dst
  }
}

object Threshold {
  val stageName: String = "threshold"
  val threshold: String = "threshold"
  val maxVal: String = "maxVal"
  val thresholdType: String = "type"
}

/** Applies gaussian kernel to blur the image. Please refer to
  * [[http://docs.opencv.org/2.4/modules/imgproc/doc/filtering.html#gaussianblur OpenCV]] for detailed information
  * about the parameters and their allowable values.
  *
  * @param params Map of parameter values containg the aperture and sigma for the kernel.
  */
class GaussianKernel(params: Map[String, Any]) extends ImageTransformerStage(params) {
  val appertureSize: Int = params(GaussianKernel.apertureSize).asInstanceOf[Int]
  val sigma: Double = params(GaussianKernel.sigma) match {
    case d: Double => d
    case i: Int => i.toDouble
  }
  override val stageName: String = GaussianKernel.stageName

  override def apply(image: Mat): Mat = {
    val dst = new Mat()
    val kernel = Imgproc.getGaussianKernel(appertureSize, sigma)
    Imgproc.filter2D(image, dst, -1, kernel)
    dst
  }
}

object GaussianKernel {
  val stageName: String = "gaussiankernel"
  val apertureSize: String = "apertureSize"
  val sigma: String = "sigma"
}

/** Pipelined image processing. */
object ImageTransformer extends DefaultParamsReadable[ImageTransformer] {

  override def load(path: String): ImageTransformer = super.load(path)

  /** Convert Spark image representation to OpenCV format. */
  private def row2mat(row: Row): (String, Mat) = {
    val path = ImageSchema.getOrigin(row)
    val height = ImageSchema.getHeight(row)
    val width = ImageSchema.getWidth(row)
    val ocvType = ImageSchema.getMode(row)
    val bytes = ImageSchema.getData(row)

    val img = new Mat(height, width, ocvType)
    img.put(0, 0, bytes)
    (path, img)
  }

  /** Convert from OpenCV format to Dataframe Row; unroll if needed. */
  private def mat2row(img: Mat, path: String = ""): Row = {
    val ocvBytes = new Array[Byte](img.total.toInt * img.elemSize.toInt)
    img.get(0, 0, ocvBytes) //extract OpenCV bytes
    Row(path, img.height, img.width, img.channels(), img.`type`, ocvBytes)
  }

  /** Apply all OpenCV transformation stages to a single image; unroll the result if needed
    * For null inputs or binary files that could not be parsed, return None.
    * Break on OpenCV errors.
    */
  def process(stages: Seq[ImageTransformerStage], decodeMode: String)(r: Any): Option[Row] = {

    if (r == null) return None

    val decoded = (r, decodeMode) match {
      case (row: Row, "binaryfile") =>
        val path = BinaryFileSchema.getPath(row)
        val bytes = BinaryFileSchema.getBytes(row)

        //early return if the image can't be decompressed
        ImageInjections.decode(path, bytes).getOrElse(return None).getStruct(0)
      case (bytes: Array[Byte], "binary") =>
        ImageInjections.decode(null, bytes).getOrElse(return None).getStruct(0)
      case (row: Row, "image") =>
        row
      case (_, mode) =>
        throw new MatchError(s"Unknown decoder mode $mode")
    }

    val (path, img) = row2mat(decoded)
    val result = stages.foldLeft(img) {
      case (imgInternal, stage) => stage.apply(imgInternal)
    }
    Some(mat2row(result, path))
  }

}

/** Image processing stage. Please refer to OpenCV for additional information
  *
  * @param uid The id of the module
  */
class ImageTransformer(val uid: String) extends Transformer
  with HasInputCol with HasOutputCol with Wrappable with DefaultParamsWritable with BasicLogging {
  logClass()

  import ImageTransformer._

  override protected lazy val pyInternalWrapper = true

  def this() = this(Identifiable.randomUID("ImageTransformer"))

  val stages: ArrayMapParam = new ArrayMapParam(this, "stages", "Image transformation stages")

  def setStages(value: Array[Map[String, Any]]): this.type = set(stages, value)

  val emptyStages = Array[Map[String, Any]]()

  def getStages: Array[Map[String, Any]] = if (isDefined(stages)) $(stages) else emptyStages

  private def addStage(stage: Map[String, Any]): this.type = set(stages, getStages :+ stage)

  setDefault(inputCol -> "image", outputCol -> (uid + "_output"))

  // every stage has a name like "resize", "normalize", "unroll"
  val stageName = "action"

  def resize(height: Int, width: Int): this.type = {
    require(width >= 0 && height >= 0, "width and height should be nonnegative")

    addStage(Map(stageName -> ResizeImage.stageName,
      ResizeImage.width -> width,
      ResizeImage.height -> height))
  }

  def crop(x: Int, y: Int, height: Int, width: Int): this.type = {
    require(x >= 0 && y >= 0 && width >= 0 && height >= 0, "crop values should be nonnegative")

    addStage(Map(stageName -> CropImage.stageName,
      CropImage.width -> width,
      CropImage.height -> height,
      CropImage.x -> x,
      CropImage.y -> y))
  }

  def colorFormat(format: Int): this.type = {
    addStage(Map(stageName -> ColorFormat.stageName, ColorFormat.format -> format))
  }

  def blur(height: Double, width: Double): this.type = {
    addStage(Map(stageName -> Blur.stageName, Blur.height -> height, Blur.width -> width))
  }

  def threshold(threshold: Double, maxVal: Double, thresholdType: Int): this.type = {
    addStage(Map(stageName -> Threshold.stageName,
      Threshold.maxVal -> maxVal,
      Threshold.threshold -> threshold,
      Threshold.thresholdType -> thresholdType))
  }

  /** Flips the image
    *
    * @param flipCode is a flag to specify how to flip the image:
    * - 0 means flipping around the x-axis (i.e. up-down)
    * - positive value (for example, 1) means flipping around y-axis (left-right)
    * - negative value (for example, -1) means flipping around both axes (diagonally)
    *                 See OpenCV documentation for details.
    * @return
    */
  def flip(flipCode: Int): this.type = {
    addStage(Map(stageName -> Flip.stageName, Flip.flipCode -> flipCode))
  }

  def gaussianKernel(apertureSize: Int, sigma: Double): this.type = {
    addStage(Map(stageName -> GaussianKernel.stageName,
      GaussianKernel.apertureSize -> apertureSize,
      GaussianKernel.sigma -> sigma))
  }

  override def transform(dataset: Dataset[_]): DataFrame = {
    logTransform[DataFrame]({

      //  load native OpenCV library on each partition
      // TODO: figure out more elegant way
      val df = OpenCVUtils.loadOpenCV(dataset.toDF)
      val decodeMode = df.schema(getInputCol).dataType match {
        case s if ImageSchemaUtils.isImage(s) => "image"
        case s if BinaryFileSchema.isBinaryFile(s) => "binaryfile"
        case s if s == BinaryType => "binary"
        case s =>
          throw new IllegalArgumentException(s"input column should have Image or BinaryFile type, got $s")

      }

      val transforms = ListBuffer[ImageTransformerStage]()
      for (stage <- getStages) {
        stage(stageName) match {
          case ResizeImage.stageName => transforms += new ResizeImage(stage)
          case CropImage.stageName => transforms += new CropImage(stage)
          case ColorFormat.stageName => transforms += new ColorFormat(stage)
          case Blur.stageName => transforms += new Blur(stage)
          case Threshold.stageName => transforms += new Threshold(stage)
          case GaussianKernel.stageName => transforms += new GaussianKernel(stage)
          case Flip.stageName => transforms += new Flip(stage)
          case unsupported: String => throw new IllegalArgumentException(s"unsupported transformation $unsupported")
        }
      }

      val convert = UDFUtils.oldUdf(process(transforms, decodeMode = decodeMode) _, ImageSchema.columnSchema)

      df.withColumn(getOutputCol, convert(df(getInputCol)))
    })
  }

  override def copy(extra: ParamMap): Transformer = defaultCopy(extra)

  override def transformSchema(schema: StructType): StructType = {
    schema.add(getOutputCol, ImageSchema.columnSchema)
  }

}

//scalastyle:on field.name

File: opencv/src/main/python/synapse/ml/opencv/ImageTransformer.py
# Copyright (C) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE in project root for information.

import sys
from typing import List

if sys.version >= "3":
    basestring = str

import pyspark
from pyspark import SparkContext
from pyspark.ml.common import inherit_doc
from pyspark.sql.types import *
from pyspark.sql.types import Row, _create_row
import numpy as np
from synapse.ml.opencv._ImageTransformer import _ImageTransformer

ImageFields = ["origin", "height", "width", "nChannels", "mode", "data"]

ImageSchema = StructType(
    [
        StructField(ImageFields[0], StringType(), True),
        StructField(ImageFields[1], IntegerType(), True),
        StructField(ImageFields[2], IntegerType(), True),
        StructField(ImageFields[3], IntegerType(), True),
        StructField(ImageFields[4], IntegerType(), True),  # OpenCV type: CV_8U in most cases
        StructField(ImageFields[5], BinaryType(), True),
    ]
)  # OpenCV bytes: row-wise BGR in most cases


def toNDArray(image):
    """
    Converts an image to a 1-dimensional array

    Args:
        image (object): The image to be converted

    Returns:
        array: The image as a 1-dimensional array
    """
    return np.asarray(image.data, dtype=np.uint8).reshape((image.height, image.width, 3))[:, :, (2, 1, 0)]


def toImage(array, path="", mode=16):
    """

    Converts a one-dimensional array to a 2 dimensional image

    Args:
        array (array):
        path (str):
        ocvType (int):

    Returns:
        object: 2 dimensional image
    """
    length = np.prod(array.shape)

    data = bytearray(array.astype(dtype=np.int8)[:, :, (2, 1, 0)].reshape(length))
    height = array.shape[0]
    width = array.shape[1]
    # Creating new Row with _create_row(), because Row(name = value, ... ) orders fields by name,
    # which conflicts with expected ImageSchema order when the new DataFrame is created by UDF
    return _create_row(ImageFields, [path, height, width, 3, mode, data])


from pyspark.ml.common import inherit_doc


@inherit_doc
class ImageTransformer(_ImageTransformer):
    """
    Transformer for common image processing stages.
    """

    def resize(self, size, keep_aspect_ratio=True):
        """
        Resizes the image to the given size.

        Args:
            size (int or tuple(width, height)): The size to resize to (>=0).
            keep_aspect_ratio (bool): Whether to keep aspect ratio.
            If true, the shorter side of the image will be resized to the specified size.
        """
        if type(size) is tuple:
            height, width = (size[1], size[0])
            self._java_obj.resize(height, width)
            return self
        else:
            self._java_obj.resize(size, keep_aspect_ratio)
            return self

    def crop(self, x, y, height, width):
        """
        Crops the image given the starting x,y coordinates
        and the width and height

        Args:
            x (int): The initial x coordinate (>=0)
            y (int): The initial y coordinate (>=0)
            height (int): The height to crop to (>=0)
            width (int): The width to crop to (>=0)

        """
        self._java_obj.crop(x, y, height, width)
        return self

    def centerCrop(self, height, width):
        """
        Center crops the image given the width and height.

        Args:
            height (int): The height to crop to (>= 0)
            width (int): The width to crop to (>= 0)

        """
        self._java_obj.centerCrop(height, width)
        return self

    def colorFormat(self, format):
        """
        Formats the image to the given image format

        Args:
            format (int): The format to convert to, please see OpenCV cvtColor function documentation for all formats

        """
        self._java_obj.colorFormat(format)
        return self

    def blur(self, height, width):
        """
        Blurs the image using a normalized box filter

        Args:
            height (double): The height of the box filter (>= 0)
            width (double): The width of the box filter (>= 0)

        """
        self._java_obj.blur(height, width)
        return self

    def threshold(self, threshold, max_val, threshold_type):
        """
        Thresholds the image, please see OpenCV threshold function documentation for more information

        Args:
            threshold: (double) The threshold value
            max_val (double): The maximum value to use
            threshold_type (double): The type of threshold, can be binary, binary_inv, trunc, zero, zero_inv

        """
        self._java_obj.threshold(threshold, max_val, threshold_type)
        return self

    def gaussianKernel(self, aperture_size, sigma):
        """
        Blurs the image by applying a gaussian kernel

        Args:
            aperture_size (double): The aperture size, which should be odd and positive
            sigma (double): The standard deviation of the gaussian

        """
        self._java_obj.gaussianKernel(aperture_size, sigma)
        return self

    def flip(self, flip_code=1):
        """
        Flips the image
        :param int flip_code: a flag to specify how to flip the image
        - 0 means flipping around the x-axis (up-down)
        - positive value (for example, 1) means flipping around y-axis (left-right, default)
        - negative value (for example, -1) means flipping around both axes (diagonally)
        See OpenCV documentation for details.
        """
        self._java_obj.flip(flip_code)
        return self

    def normalize(self, mean, std, color_scale_factor):
        """
        Normalizes the image by multiplying the color_scale_factor, substracting mean and dividing by std
        """
        self._java_obj.normalize(mean, std, color_scale_factor)
        return self
